### ISSUE TITLE
Breaking behaviour change in error handling between 1.11.6 and 1.11.7

### ISSUE THREAD

**Issue by Bios-Marcel** [twmb/franz-go#357]
This was caused by the change in `pkg/kgo/source.go` regarding the error treatment.

Before, I was able to react to certain errors, reconnect and receive messages again.

If you run the below example with `v1.11.6`, it will print that it tries to reconnect, which in my real codebase worked and led to consuming the messages that were publish after reseting to a fresh kafka instance.

With `v1.11.7`, this behaves differently. We wait for the messages forever and no attempt to reconnect is possible.

It doesn't seem to matter whether a group has been specified or not.
Either way, behaviour changed :thinking: 

The example starts a kafka instance, resets it and purges it again.
The example should run as long as you have docker installed.
Its quite big, but only the main method really matters i guess.

```go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"os"
	"sync"
	"time"

	"github.com/gofrs/uuid"
	"github.com/ory/dockertest/v3"
	"github.com/ory/dockertest/v3/docker"
	"github.com/twmb/franz-go/pkg/kerr"
	"github.com/twmb/franz-go/pkg/kgo"
)

func main() {
	assertNoError(kafka.reset(), "error starting kafka")
	defer func() {
		// If it is error unsupported, it means the tests didn't require kafka.
		if err := kafka.purge(); err != nil && err != errUnsupported {
			fmt.Printf("Coludn't purge Kafka: %s\n", err)
		}
	}()

	produceClient, err := kgo.NewClient(
		kgo.SeedBrokers("localhost:9092"),
		kgo.AllowAutoTopicCreation(),
	)
	if err != nil {
		panic(err)
	}

	topic := "reconnect-topic"
	group := "reconnect-group"

	listenClient, err := kgo.NewClient(
		kgo.SeedBrokers("localhost:9092"),
		kgo.ConsumeTopics(topic),
		kgo.ConsumerGroup(group),
	)
	if err != nil {
		panic(err)
	}

	var wg sync.WaitGroup
	log.Println("Listening ...")
	go func() {
		defer log.Println("Listener has closed")
		for {
			fetches := listenClient.PollRecords(context.Background(), 1)
			if fetches.IsClientClosed() {
				// No intention of retrying, as the listener closed has no more
				// interest in listening.
				log.Println("Client Closed")
				os.Exit(1)
				return
			}

			if kgoErrors := fetches.Errors(); len(kgoErrors) > 0 {
				for _, kgoError := range kgoErrors {
					// Documentation for 1.11.6 indicates we might want to restart listening here.
					if err, ok := kgoError.Err.(*kerr.Error); ok {
						log.Println("Imagine reconnect here (We expect this)", err)
						os.Exit(0)
					}

					log.Println("error", kgoError)
				}

				log.Println("Imagine listener invocation here", kgoErrors)
				wg.Done()
				continue
			}

			// We can iterate through a record iterator...
			iter := fetches.RecordIter()
			for !iter.Done() {
				record := iter.Next()
				log.Println("Imagine listener invocation here", string(record.Value))
				wg.Done()
			}
		}
	}()

	log.Println("Publishing ...")
	wg.Add(2)
	assertNoError(publish(produceClient, topic, map[string]string{"field": "a"}), "error publishing message")
	assertNoError(publish(produceClient, topic, map[string]string{"field": "b"}), "error publishing message")

	log.Println("Waiting to receive messages ...")
	wg.Wait()

	log.Println("Shutting down kafka for reconnect test ...")
	assertNoError(kafka.reset(), "error restarting kafka")

	log.Println("Publishing ...")
	wg.Add(2)
	assertNoError(publish(produceClient, topic, map[string]string{"field": "c"}), "error publishing message")
	assertNoError(publish(produceClient, topic, map[string]string{"field": "d"}), "error publishing message")

	log.Println("Waiting to receive messages ...")
	wg.Wait()
}

func assertNoError(err error, msg string) {
	if err != nil {
		panic(fmt.Errorf(msg+":%w", err))
	}
}

func publish(client *kgo.Client, topic string, payload any) error {
	bytes, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	record := kgo.SliceRecord(bytes)
	record.Topic = topic
	result := client.ProduceSync(context.Background(), record)
	return result.FirstErr()
}

type kafkaInstance struct {
	host string
	port int

	start func() error
	// _purge must not be called directly, call purge instead.
	_purge func() error
}

var errUnsupported = errors.New("unsupported operation")

var kafka = &kafkaInstance{
	host: "localhost",
	port: 9092,
	start: func() error {
		return errUnsupported
	},
	// Initially nil, cause there isn't anything to purge
	_purge: nil,
}

func (k *kafkaInstance) purge() error {
	fmt.Printf("Purging Kafka instance at %s:%d\n", k.host, k.port)
	if k._purge != nil {
		return k._purge()
	}

	return nil
}

func (k *kafkaInstance) wait() {
	address := fmt.Sprintf("%s:%d", k.host, k.port)
	timeout := 30 * time.Second
	fmt.Printf("Waiting %s for Kafka at %s ...\n", timeout.String(), address)
	if !available(address, timeout) {
		fmt.Println("Kafka not found at", address)
		os.Exit(1)
	}

	fmt.Println("Kafka available at", address)
}

func (k *kafkaInstance) available() bool {
	address := fmt.Sprintf("%s:%d", k.host, k.port)
	fmt.Printf("Check for Kafka at %s ...\n", address)
	if !availableNoWait(address) {
		fmt.Println("Kafka not found at", address)
		return false
	}

	fmt.Println("Kafka available at", address)
	return true
}

func (k *kafkaInstance) reset() error {
	fmt.Printf("Resetting Kafka instance at %s:%d\n", k.host, k.port)
	if err := k.purge(); err != nil {
		return err
	}

	if err := k.start(); err != nil {
		return err
	}

	return nil
}

// availableNoWait does the same check as Available, but only attempts it once
// and instantly (however long one try takes) returns, no matter the result.
func availableNoWait(address string) bool {
	options := []kgo.Opt{
		kgo.SeedBrokers(address),
		// AutoTopicCreation is disabled by default.
	}
	client, err := kgo.NewClient(options...)
	if err != nil {
		return false
	}
	record := kgo.StringRecord("ANYONE THERE?")
	record.Topic = "non-existent" + uuid.Must(uuid.NewV4()).String()

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	result := client.ProduceSync(ctx, record)
	if err := result.FirstErr(); err != nil && !errors.Is(err, kerr.UnknownTopicOrPartition) {
		return false
	}
	return true
}

// Available attempts to send a message to a non-existent topic. Autocreation
// of the topic is disable during this process, in order to make sure we
// receive a "topic doesn't exist" type of error. This message blocks until
// we know that Kafka is available or run into the timeout.
func available(address string, timeout time.Duration) bool {
	options := []kgo.Opt{
		kgo.SeedBrokers(address),
		// AutoTopicCreation is disabled by default.
	}
	client, err := kgo.NewClient(options...)
	if err != nil {
		return false
	}

	timeoutTimer := time.NewTimer(timeout)
	resultChannel := make(chan bool, 1)

	go func() {
		record := kgo.StringRecord("ANYONE THERE?")
		record.Topic = "non-existent" + uuid.Must(uuid.NewV4()).String()

		// Retry till the select runs into the timeout.
	RETRY:
		result := client.ProduceSync(context.Background(), record)
		err := result.FirstErr()
		if err != nil && !errors.Is(err, kerr.UnknownTopicOrPartition) {
			goto RETRY
		}
		resultChannel <- true
	}()

	select {
	case <-timeoutTimer.C:
		return false
	case resultValue := <-resultChannel:
		client.Close()
		timeoutTimer.Stop()
		return resultValue
	}
}

func init() {
	kafka.host = "localhost"
	kafka.port = 9092
	kafka.start = func() error {
		// Prevent trying to start multiple times, as we'll run into an error.
		if kafka.available() {
			fmt.Printf("Not starting Kafka on port %d, as it is already available\n", kafka.port)

			// Since no purge is available yet, we assume that we haven't
			// started this instance and therefore can't purge it.
			if kafka._purge == nil {
				kafka._purge = func() error {
					return errUnsupported
				}
			}
			return nil
		}

		fmt.Printf("Starting Kafka on port %d ...\n", kafka.port)

		// Connects to the local docker daemon.
		pool, err := dockertest.NewPool("")
		if err != nil {
			fmt.Printf("Could not connect to docker: %s\n", err)
			return err
		}

		// Cleanup container after 120 seconds in case our cleanup fails.
		// I went for 120 seconds in the assumption that the tests will
		// never be this slow ;)
		var expireTime uint = 120

		portString := fmt.Sprintf("%d/tcp", kafka.port)
		kafkaResource, err := pool.RunWithOptions(
			&dockertest.RunOptions{
				Repository: "blacktop/kafka",
				Tag:        "latest",
				Env: []string{
					"KAFKA_ADVERTISED_HOST_NAME=localhost",
				},
				PortBindings: map[docker.Port][]docker.PortBinding{
					"9092/tcp": {{HostIP: "localhost", HostPort: portString}},
				},
				ExposedPorts: []string{portString},
			})
		if err != nil {
			fmt.Printf("Could not start kafka resource: %s\n", err)
			return err
		}
		kafkaResource.Expire(expireTime)

		kafka.wait()
		kafka._purge = func() error {
			// You can't defer this because os.Exit doesn't care for defer
			if err := pool.Purge(kafkaResource); err != nil {
				return err
			}
			return nil
		}
		return nil
	}
}

```

**Comment by twmb** [comment]
Are you expecting to see `UNKNOWN_TOPIC_ID`? This is what I get on v1.11.6 for the code above (also please include the logs you're seeing before/after for future issues so I don't have to run this myself :sweat_smile:)

**Comment by Bios-Marcel** [comment]
Thanks.

Sorry, I thought you wanted to run it yourself, so you see what I mean.
Next time :D

I guess my usecase is rather special though. As we have an instance that is purged and created again. In most usecases you'd probably want to retain your topics, in which case that error probably makes sense.

Is there some way we could add helper code to make sure people can easily allow franz-go to retry where sensible, but opt-into handling it themselves?

**Comment by twmb** [comment]
The fix is is in v1.12.1, released now.a

**Comment by twmb** [comment]
Also to answer your question:

> Is there some way we could add helper code to make sure people can easily allow franz-go to retry where sensible, but opt-into handling it themselves?

I think it makes sense to fail on a topic recreation and bubble the error up. That's part of the reason topic IDs were introduced, to detect and guard against recreations. They just made the error retryable for some reason. Instead, I've created #358 to add an option to allow consuming recreated topics.
